### PR TITLE
Use storage access framework for backing up the wallet.

### PR DESCRIPTION
**Pull request by schildbach**
This gets rid of the WRITE_EXTERNAL_STORAGE permission. Users can still save to external storage via the SAF dialog, but actually this is discouraged as almost no devices have SD cards today. Much better to use cloud storage these days (e.g. Google Drive).

It should be noted that saving to email attachment is not supported any more either.

Note also that restoring wallets didn't change, so you can still restore from all old locations, including Google Drive and SD cards. To change this is planned for a future PR.

### CODE REVIEW

**Comment by schildbach** [comment]
@HashEngineering I wonder what you think of this change?

File: wallet/res/values/strings.xml [hunk 1]
```diff
@@ -204,18 +204,12 @@
...
     <string name="backup_wallet_dialog_warning_encrypted">Your wallet is protected by a spending PIN. Make sure you remember the PIN in addition to the backup password!</string>
     <string name="export_keys_dialog_button_export">Back up</string>
-    <string name="export_keys_dialog_success"><![CDATA[<p>Your wallet has been backed up to <tt>%s</tt></p><p>If your device has a removable SD card, your backup <i>should</i> be on that card. Remove it and keep it at a safe place. If the card is <i>not</i> removable, you need to archive your backup to some other place than your device.</p><p><b>If the only place your backup exists is on your device, you run the risk of losing both at the same time!</b></p><p>In any case, make sure you remember your backup password. Do you want to archive your wallet now?</p>]]></string>
+    <string name="export_keys_dialog_success"><![CDATA[<p>Your wallet has been backed up to <tt>%s</tt></p><p><b>If the only place your backup exists is on your device, you run the risk of losing both at the same time!</b></p><p>In any case, make sure you remember your backup password.</p>]]></string>
```
Comment (HashEngineering) [inline wallet/res/values/strings.xml]: This part can be confusing for the user.  The location will be displayed as a URI that won't connect in the user's mind to a Google Drive folder.

Secondly, there is the statement "If the only place your backup exists is on your device...", but user just put the backup on a cloud.

screenshot_20180208-001935

**Comment by HashEngineering** [comment]
I built this on my system and looked at how it works.

I think it is a good idea.  It takes the backup process and makes it more simple and more visual to the user as to where the backup file is going.  Many users have asked how to restore their wallets because they don't know where the backup file is, that it was stored in the /Download folder, how to find it when restoring, what archiving means, etc.  

I have made one comment about the dialog that pops up after Backup Process is complete regarding the text about the storage location and if the backup is on the same device.

I look forward to your other PR for restoring backups.

**Comment by schildbach** [comment]
I agree about the storage location. Sadly I could not find a way to find out what location the user has picked in a more user understandable way. I propose to replace URLs starting with `content://com.google.android.apps.docs.storage` by `Google Drive` and `content://com.android.providers.downloads.documents` with something like `local storage`. We can add more providers later as they hopefully start to support the SAF (Dropbox, do you read this?)

**Comment by schildbach** [comment]
I left the local storage warning in, because it *is* still possible to save to local storage (although hidden somewhat in the options menu of the SAF dialog). Would you suggest to just remove the warning?

**Comment by HashEngineering** [comment]
I had not noticed that the user could also save to local storage using this method.  Therefore the warning should stay.

**Comment by schildbach** [comment]
I implemented the URI replacement as proposed above, using the term `internal storage` rather than `local storage` as that term is used in the Google Drive dialog too.

**Pull request merged**